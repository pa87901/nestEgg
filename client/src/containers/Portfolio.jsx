import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Segment, Button } from 'semantic-ui-react';
import 'isomorphic-fetch';
import Blotter from '../components/Blotter';
import { setHoldings, removeBookings } from '../actions/holdingActions';


class Portfolio extends Component {
  constructor() {
    super();
    this.delete = this.delete.bind(this);
  }

  componentWillMount() {
    const { handleSetHoldings } = this.props;
    fetch('http://localhost:3000/api/holdings', { method: 'get' })
    .then(holdings => holdings.json())
    .then(holdingsJSON => {
      handleSetHoldings(holdingsJSON);
    })
    .catch(err => {
      console.error('Unable to get all holdings from the backend:', err); // eslint-disable-line no-console
    });
  }

  delete() {
    const { selectedHoldings, handleRemoveBookings } = this.props;
    if (!selectedHoldings.length) {
      window.alert('No holdings selected. Please select holdings by checkboxes.');
      return;
    }
    console.log('selectedHoldings:', selectedHoldings);
    const payload = { selectedHoldings };
    const headers = {
      'accept': 'application/json, text/plain, */*',
      'content-type': 'application/json'
    };
    const init = {
      method: 'DELETE',
      headers,
      mode: 'cors',
      body: JSON.stringify(payload),
      json: true
    };
    fetch('/api/holdings/', init)
    .then(res => res.json())
    .then(resJSON => {
      // Use this for MongoDB database
      handleRemoveBookings(resJSON.selectedHoldings);
      /* Use this for PSQL database
      handleRemoveBookings(resJSON.rows);
      */
    })
    .catch(err => {
      console.error('Error sending ids to delete:', err); // eslint-disable-line no-console
    });
  }

  render() {
    const { selectedHoldings } = this.props;
    return (
      <div>
        <Segment.Group className="blotter">
          <Segment>
            <h2>Holdings</h2>
            <Blotter />
          </Segment>
          <Segment className="portfolioButtonRow" >
            <Button
              basic
              color={selectedHoldings.length ? "red" : null}
              onClick={this.delete}>
                Delete
            </Button>
          </Segment>
        </Segment.Group>
      </div>
    );
  }
}

Portfolio.propTypes = {
  handleSetHoldings: PropTypes.func.isRequired,
  selectedHoldings: PropTypes.array.isRequired, // eslint-disable-line react/forbid-prop-types
  handleRemoveBookings: PropTypes.func.isRequired
}

const mapStateToProps = state => ({
  selectedHoldings: state.holdings.selectedHoldings
});

const mapDispatchToProps = dispatch => ({
  handleSetHoldings(holdings) {
    dispatch(setHoldings(holdings))
  },
  handleRemoveBookings(holdings) {
    dispatch(removeBookings(holdings));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Portfolio);