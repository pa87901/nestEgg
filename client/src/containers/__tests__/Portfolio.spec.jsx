import React from 'react';
import { shallow } from 'enzyme';
import { Unwrapped as UnwrappedPortfolio } from '../Portfolio';

describe('Portfolio', () => {
  const shallowComponent = shallow(
    <UnwrappedPortfolio
      store={[]}
      selectedHoldings={[]}
      handleSetHoldings={() => {}}
      handleRemoveBookings={() => {}} />
    );
  test('Portfolio snapshot has not changed', () => {
    expect(shallowComponent).toMatchSnapshot();
  });
});