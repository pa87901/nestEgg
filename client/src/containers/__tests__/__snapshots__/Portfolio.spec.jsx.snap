// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portfolio Portfolio snapshot has not changed 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Portfolio
    handleRemoveBookings={[Function]}
    handleSetHoldings={[Function]}
    selectedHoldings={Array []}
    store={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <SegmentGroup
        className="blotter"
      >
        <Segment>
          <h2>
            Holdings
          </h2>
          <Connect(Blotter) />
        </Segment>
        <Segment
          className="portfolioButtonRow"
        >
          <Button
            as="button"
            basic={true}
            color={null}
            onClick={[Function]}
          >
            Delete
          </Button>
        </Segment>
      </SegmentGroup>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Segment>
            <h2>
              Holdings
            </h2>
            <Connect(Blotter) />
          </Segment>,
          <Segment
            className="portfolioButtonRow"
          >
            <Button
              as="button"
              basic={true}
              color={null}
              onClick={[Function]}
            >
              Delete
            </Button>
          </Segment>,
        ],
        "className": "blotter",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <h2>
                Holdings
              </h2>,
              <Connect(Blotter) />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Holdings",
              },
              "ref": null,
              "rendered": "Holdings",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Button
              as="button"
              basic={true}
              color={null}
              onClick={[Function]}
            >
              Delete
            </Button>,
            "className": "portfolioButtonRow",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "button",
              "basic": true,
              "children": "Delete",
              "color": null,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Delete",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <SegmentGroup
          className="blotter"
        >
          <Segment>
            <h2>
              Holdings
            </h2>
            <Connect(Blotter) />
          </Segment>
          <Segment
            className="portfolioButtonRow"
          >
            <Button
              as="button"
              basic={true}
              color={null}
              onClick={[Function]}
            >
              Delete
            </Button>
          </Segment>
        </SegmentGroup>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Segment>
              <h2>
                Holdings
              </h2>
              <Connect(Blotter) />
            </Segment>,
            <Segment
              className="portfolioButtonRow"
            >
              <Button
                as="button"
                basic={true}
                color={null}
                onClick={[Function]}
              >
                Delete
              </Button>
            </Segment>,
          ],
          "className": "blotter",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <h2>
                  Holdings
                </h2>,
                <Connect(Blotter) />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Holdings",
                },
                "ref": null,
                "rendered": "Holdings",
                "type": "h2",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Button
                as="button"
                basic={true}
                color={null}
                onClick={[Function]}
              >
                Delete
              </Button>,
              "className": "portfolioButtonRow",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "basic": true,
                "children": "Delete",
                "color": null,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Delete",
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
